### PR TITLE
+per #16339 adds actorSelection support to AtLeastOnceDelivery

### PR DESCRIPTION
**Pull request by ktoso**
Resolves https://github.com/akka/akka/issues/16339

### CODE REVIEW

File: akka-persistence/src/main/scala/akka/persistence/AtLeastOnceDelivery.scala [hunk 1]
```diff
@@ -197,7 +196,7 @@ trait AtLeastOnceDelivery extends Eventsourced {
    * This method will throw [[AtLeastOnceDelivery.MaxUnconfirmedMessagesExceededException]]
    * if [[numberOfUnconfirmed]] is greater than or equal to [[#maxUnconfirmedMessages]].
    */
-  def deliver(destination: ActorPath, deliveryIdToMessage: Long ⇒ Any): Unit = {
+  def deliver(destination: ActorPath)(deliveryIdToMessage: Long ⇒ Any): Unit = {
```
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/AtLeastOnceDelivery.scala]: Without splitting it up in 2 parameter lists inference breaks and is unable to figure out that `{ id => }` is `Long`.

**Comment by ktoso** [comment]
While I'm amending docs, would you guys give me a hand and review this if you have time? :)
// @2m @viktorklang

**Comment by ktoso** [comment]
Docs amended to show selection instead of actor path; both remain available though

**Comment by ktoso** [comment]
Added migration guide, there's a minor breakage in `deliver()()` (OK, since was experimental)

**Comment by jrudolph** [comment]
Error lines in https://jenkins.akka.io/job/pr-validator-per-commit-jenkins/3272/:

```
  1968 /mnt/akka/jenkinsakka/localhome-a0/workspace/pr-validator-per-commit-jenkins/akka-docs/rst/java/code/docs/persistence/LambdaPersistenceDocTest.java:173: error: cannot find symbol
  1969         private final ActorSelection destination;
  1970                       ^
  1971   symbol:   class ActorSelection
  1972   location: class <anonymous Object>.MyPersistentActor
  1973 /mnt/akka/jenkinsakka/localhome-a0/workspace/pr-validator-per-commit-jenkins/akka-docs/rst/java/code/docs/persistence/LambdaPersistenceDocTest.java:175: error: cannot find symbol
  1974         public MyPersistentActor(ActorSelection destination) {
  1975                                  ^
  1976   symbol:   class ActorSelection
  1977   location: class <anonymous Object>.MyPersistentActor
  1978 /mnt/akka/jenkinsakka/localhome-a0/workspace/pr-validator-per-commit-jenkins/akka-docs/rst/java/code/docs/persistence/PersistenceDocTest.java:150: error: cannot find symbol
  1979         private final ActorSelection destination;
  1980                       ^
  1981   symbol:   class ActorSelection
  1982   location: class <anonymous Object>.MyPersistentActor
  1983 /mnt/akka/jenkinsakka/localhome-a0/workspace/pr-validator-per-commit-jenkins/akka-docs/rst/java/code/docs/persistence/PersistenceDocTest.java:155: error: cannot find symbol
  1984         public MyPersistentActor(ActorSelection destination) {
  1985                                  ^
  1986   symbol:   class ActorSelection
  1987   location: class <anonymous Object>.MyPersistentActor
  1988 Note: /mnt/akka/jenkinsakka/localhome-a0/workspace/pr-validator-per-commit-jenkins/akka-docs/rst/java/code/docs/actorlambda/ActorDocTest.java uses or overrides a deprecated API.
  1989 Note: Recompile with -Xlint:deprecation for details.
  1990 4 errors
  2011 (akka-docs/test:compile) javac returned nonzero exit code
  2012 Total time: 858 s, completed Jul 22, 2015 5:37:55 PM
```

File: akka-actor/src/main/scala/akka/util/Unsafe.java [hunk 2]
```diff
@@ -1,13 +1,30 @@
...
+                    found = (sun.misc.Unsafe) field.get(null);
+                    break;
+                }
+            }
+            if (found == null) throw new IllegalStateException("Can't find instance of sun.misc.Unsafe");
+            else instance = found;
+        } catch (Throwable t) {
+            throw new ExceptionInInitializerError(t);
+        }
+    }
```
Comment (patriknw) [inline akka-actor/src/main/scala/akka/util/Unsafe.java]: how is this related to the AtLeastOnceDelivery issue? separate PR?

File: akka-persistence/src/main/scala/akka/persistence/AtLeastOnceDelivery.scala [hunk 3]
```diff
@@ -213,6 +212,35 @@ trait AtLeastOnceDelivery extends Eventsourced {
...
+   * translation is performed.
+   *
+   * During recovery this method will not send out the message, but it will be sent
+   * later if no matching `confirmDelivery` was performed.
+   *
+   * This method will throw [[AtLeastOnceDelivery.MaxUnconfirmedMessagesExceededException]]
+   * if [[numberOfUnconfirmed]] is greater than or equal to [[#maxUnconfirmedMessages]].
+   */
+  def deliver(destination: ActorSelection)(deliveryIdToMessage: Long ⇒ Any): Unit = {
+    // TODO we may eventually support this, however it adds quite some complexitly to ACK tracking for users
```
Comment (patriknw) [inline akka-persistence/src/main/scala/akka/persistence/AtLeastOnceDelivery.scala]: I don't think we will. That would be very difficult (impossible). Remove the TODO if you agree.

File: akka-persistence/src/test/scala/akka/persistence/AtLeastOnceDeliverySpec.scala [hunk 4]
```diff
@@ -48,9 +52,13 @@ object AtLeastOnceDeliverySpec {
     var lastSnapshotAskedForBy: Option[ActorRef] = None
 
     def updateState(evt: Evt): Unit = evt match {
+      case AcceptedReq(payload, destination) if actorSelectionDelivery ⇒
+        log.debug(s"deliver(destination, deliveryId ⇒ Action(deliveryId, $payload)), recovery: " + recoveryRunning)
```
Comment (patriknw) [inline akka-persistence/src/test/scala/akka/persistence/AtLeastOnceDeliverySpec.scala]: we should avoid `toString` of arbitrary messages (payload), it might be huge and fill up memory and disk

Comment (patriknw) [inline akka-persistence/src/test/scala/akka/persistence/AtLeastOnceDeliverySpec.scala]: sorry, didn't see that this was a test

**Comment by patriknw** [comment]
Refs #16339

**Comment by patriknw** [comment]
LGTM, after comments

File: akka-actor/src/main/scala/akka/util/Unsafe.java (hunk 2 above)
Comment (ktoso) [inline akka-actor/src/main/scala/akka/util/Unsafe.java]: whoop, good catch, not sure how this commit ended up here.

File: akka-persistence/src/main/scala/akka/persistence/AtLeastOnceDelivery.scala (hunk 3 above)
Comment (ktoso) [inline akka-persistence/src/main/scala/akka/persistence/AtLeastOnceDelivery.scala]: I agree, let's not; removing todo

File: akka-persistence/src/test/scala/akka/persistence/AtLeastOnceDeliverySpec.scala (hunk 4 above)
Comment (ktoso) [inline akka-persistence/src/test/scala/akka/persistence/AtLeastOnceDeliverySpec.scala]: Thanks for watching out for those anyway :)

File: akka-persistence/src/main/scala/akka/persistence/AtLeastOnceDelivery.scala (hunk 3 above)
Comment (drewhk) [inline akka-persistence/src/main/scala/akka/persistence/AtLeastOnceDelivery.scala]: +1

**Comment by drewhk** [comment]
LGTM
